cpp/memilio/math/interpolation.h

```diff
@@ -43,12 +46,44 @@ auto linear_interpolation(const X& x_eval, const X& x_1, const X& x_2, const V&
     return y1 + weight * (y2 - y1);
 }
 
+/**
+ * @brief Linear interpolation of a TimeSeries.
+ * Assumes that the time points are monotonic increasing. If the time series is strictly monotonic, this function is
+ * continuous in time.
+ * If time is outside of the provided time points, this function has a constant value of the first/last time point.
+ * @param[in] time The time at which to evaluate.
+ * @param[in] data Time points to interpolate. At least one is required.
+ * @return Interpolation result.
+ */
+template <class FP>
+typename TimeSeries<FP>::Vector linear_interpolation(FP time, const TimeSeries<FP>& data)
+{
+    assert(data.get_num_time_points() > 0 && "Interpolation requires at least one time point.");
+    auto tp_range = data.get_times();
+    // find next time point in data (strictly) after time
+    const auto next_tp = std::upper_bound(tp_range.begin(), tp_range.end(), time, [](auto&& t, auto&& tp) {
+        return t < tp;
+    });
+    // interpolate in between values if possible, otherwise return first/last value
+    if (next_tp == tp_range.begin()) { // time is before first data point
+        return data.get_value(0);
+    }
+    else if (next_tp == tp_range.end()) { // time is past last data point
+        return data.get_last_value();
+    }
+    else { // time is in between data points
+        const auto i = next_tp - tp_range.begin();
+        return linear_interpolation(time, data.get_time(i - 1), data.get_time(i), data.get_value(i - 1),
+                                    data.get_value(i));
+    }
+}
+

```

Comment:
can you be more specific about what's happening? 

I don't see a mistake and this (slightly adapted version) seems to work: [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGISVykrgAyeAyYAHI%2BAEaYxP5mpAAOqAqETgwe3r7%2BgSlpjgKh4VEssfGSiXaYDhlCBEzEBFk%2BfgG2mPaFDPWNBMWRMXEJtg1NLTntCmP9YYNlw5UAlLaoXsTI7BzmAMxhyN5YANTmZth4wIwA9B7EmKcmGgCCu/uHmCc7bgBuNUTEJh22EeLzMewYBy8x0BblECiUTUBwOerwh70%2BbjEwBIhAQLCRINRkOhXyc0zurAJKOeBEwLCSBlpGIIAE8koxWB8ABqkI6s9nMNhHABqVKeTC8RCOtDmjQA%2BmFacQUoyMhA0AxpkcueYAGxHVRyzDfMS8jVanVmfWGwJHc0EbV6g1yxJ2gRa0VWo4s232kVOllmJaEgDsVmeR0jbs1DolUoA7pgLggHYCACJHCCG41iT5p9POrgnK1Om15guGoOfABijrDGhMIYzIEz2ZNtCOAFpC0sjldWy6uz3AeGnlGjncCOsGN6i9Yjonkw6AFSZwNDn3Bnajxv56lPWn0xkfGEHJjwo7V5RiuOoaWy4gKwRxFVMboQK98vBsM3uh3TdAQBAX4HBIGFzkuPwQAAJUwKhwIuRggOFP4SC5dACSBJ10DfJhfxjI4AKAkD/hhdA1miegCS9Rw2AUYMUTDEFxy8NIjCOfAqCoOV%2BRPHYM14wUPiI4DUIBL5yK8Sj7iBIDOKoOJGE2Hi2Rk0dx3PBECAgWjMAUAA6NIAC9MAgXskSODRi11J1TgASWfZVPDfDIJ0wABHLw8DuBQjjfaVMHPB0BA%2BXSjhSRV9NOLd1KjP1bydAgkjlYhDEuPMvzokdmKjK5%2BxoBh0COcJVAdMKIsEI4wg43DM3JPAHFoFleyYKglUy%2B4Iziv8/MlO8SoIHiknHAsRK8JJ2UfaI1kKnTktSoxMH02JgDCMzeSSlK0qW1x1o6mKcsjP15JU9kOM6BoMoGoah02hbLmWzBVoYMyRz7fsmAAayYPkEA%2BMIsFUI5UCoX6PnGyajmmrxCsO97CMwB1aFQRNiD8wqjghuIOvC1BFSqwrMFUOG8qOFgAbBu1zz0hc/ruWnFKOJMCDp4qicGpL03TO7tse56zOBtHrs5/jufm3ndoYsdcv7CBppZqmGAYVAHViPliC8D5/JZ4TOQ63lohZSnNLibpfv8nCGn0y5BoYHweO/TA5QqggFAFiyNCl8djrwLjTo%2BZHUYykSWCYVQICwWhLu7W0I997jeKWT3sq6o6epO3jMYm7HRoIQCQHJl7I4aXk479xPdIM4zTN7GODtT%2BHFRfZymViAhE0YHGXYMtyp2ITVKZoYgtRIaUgqOdtNaq0HA7iLn%2BKxgEG8nad73CeUm6c1UBAgESz3pHTHY2x2FBMABWCxZ4BM%2B02PrKL8X8%2B0xWDrT4vq%2Bn9IOHxx/3%2B/9/yu58LCPxvryS2TAgEf1ATVBoQCQHPxTi8JshJnj41DmtKWjZYqRhEiRUelcMpYMsvpDQvIuAkJOMg7ccNbwTznPxI4EEkIgBQqBYg6EIBcHrtLSMtDvhmAykwqCrD/gcK4Yg8c3x6FuBhMQhs1CG78IxLI8h8jsGETzsRMSCFIJAVgvBL4QjkJiXQphbAMCfpEKkbyJRu4JFRloZvYaBYZTr0fE4luaoND6TPmA3Cd89LcPHHvNYqYviyKcUAhsN9lFfA0fnVwtBEG7g4CsWgnAz68D8NwXgqBOAyMsPOBQawNgnjBDwUgBBNCpJWH9JgWB4jrU%2BiAM%2BOx9JcB2AATl1AADl1L0yQGgQw7C4IEdJHBJC8BYBIDQpDslaFIHkjgvAFAgFIVUjgWgVhwFgEgUJSRJTkEoGgekdA4gRE5JwVQfTOy6kkEcYAyBkBHFGfpHYvBMD4H%2BHgQCgR%2BCCBEGIdgUgZCCEUCodQmydB6HjKlJInAeBpIyVk6pizOAAHlJQHOCqDa5upbn3Mec8159yIAeFOfQNGuwuG8A2Vs0gEAkAnKSGcsgFB1SoApcMYAozpA0CjnEVZctUXRDCI0FkCLeCiuYMQFk6LojaD%2BJK0gJy2CCHRQwJqqKsDRC8MATEtBaCrJyaQLAocjDiChaa7yfw8C/GNQsomNRJRbAqU3cZCyZTRFSrKjwWBUUEGIN%2BZVvxiDTSUGmOkaVXGgChSsKgBhgAKGFHgTA8Z0UCmVf84QohxAguzeCtQqLdBjLSigaw1h9B4GiKs2AQkQBJkgqQUNnAdg7BDDUxZSRujGs7ABdMphCmWC4CGLsAB1MQHZOxjvZqlcdcRiCj07OyHCggGqdgXSQBQuTQ1BqwLW9a1RajOAgK4CYfhAghDmKUcoeh8jpAEOeu9qQH0MAGDe4YgQj3dF6OMTwrQ9DfrqDMd9Qx4hfpmE%2BiDfRQMLHAysYp6xNgSCRRwTJpB5m5M4AaG5dyHlPJeeQ0luBCCj2pUsWl1TtlICdcgPqbLKCNCTcoQwnQhAIBRtkipzK6AuQECx8ItB2OcdRTxylIAeWVBVZylllKLlsE4LwMTcRMXCA4/GTDQRVA1CeMQJNimtM1HqPgbJvBs2ArzXy2QhbIULJLfoMtg6rCWCrTW%2BAEB62NsYM24YHAdiDM7agbtGRe39v4k56wI7x2TvnaVOd07N1o2XZ8wwjhkAbuIIu4eXZ0VmB3Qun5el4ArA1mutgAAVVAngD0IZKchiDJmBNsfU1x3gsKmDwpyah9DmmlnYG03R/4OH8V4aJS8yQ%2BkBFkorS5xhXyyPlIo5UqjtTAoNMoCsZpkhdT6T6RoEZbauBn06VaHpnT9CcEmaQaZXBZkYdRUslZIBEh0s7eMvL12ZlzIewZ17SwVihrSM4SQQA%3D)

Reply:
The problem seems to be the operator[] on a temporary. In the TimeSeriesFunctor using it results in garbage output, but using .coeff instead works as intended:
```cpp
        case TimeSeriesFunctorType::LinearInterpolation:
            // return linear_interpolation(time, m_data)[0];
            return linear_interpolation(time, m_data).coeff(0);
```

Reply:
I intend to just use coeff here, but if you are interested in some more detail:

Inside `linear_interpolation(time, m_data)` the calculation works just fine, and its result is a (lazy evaluation of a) single element vector with the correct value. But, when trying to use its value in the TimeSeriesFunctor via operator[] or cout<<, it appears as an uninitialized vector, i.e. random size and values - if used as `auto x = linear_interpolation(time, m_data)[0];`. Interestingly, when returned directly instead, it shows deterministic values for the test TestMathTimeSeriesFunctor.linearInterpolation, that are only wrong on the slope (i.e. for t=0.25, 0.5, 0.75):

```
The difference between tsf(t) and pcw_lin_fct(t) is 0.75, which exceeds 1e-14, where
tsf(t) evaluates to 0.75,
pcw_lin_fct(t) evaluates to 1.5, and
1e-14 evaluates to 1e-14.

The difference between tsf(t) and pcw_lin_fct(t) is 1.5, which exceeds 1e-14, where
tsf(t) evaluates to 0.5,
pcw_lin_fct(t) evaluates to 2, and
1e-14 evaluates to 1e-14.

The difference between tsf(t) and pcw_lin_fct(t) is 2.25, which exceeds 1e-14, where
tsf(t) evaluates to 0.25,
pcw_lin_fct(t) evaluates to 2.5, and
1e-14 evaluates to 1e-14.
```

The wrong values appear to be `1 - t`. I don't know where they could come from, though.

Reply:
that's weird, coeff and operator should be the same except for one assert. maybe an effect of compiler optimizations? 

this [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGEgBykrgAyeAyYAHI%2BAEaYxCBmXKQADqgKhE4MHt6%2BASlpGQKh4VEssfGJtpj2jgJCBEzEBNk%2BflyBdpgOmfWNBMWRMXEJSQoNTS257bbj/WGDZcOJAJS2qF7EyOwc5gDMYcjeWADU5mbYeMCMAPQexJhnJhoAgnsHR5inu24Abl1ExBMu2wT1eZn2DEOXhOQLcogUSiaQJBLzekI%2BXzcYmAJEICBYyNBaKhMO%2BTjG91YhNRLwpXgcxwAKkIiQB2KwvY5c44XK5%2BEAAWSYBGIeFUAA10McCbsOa9OdzmRBlqDuccQNKIGZSMczCqFVyTOzVWruSwIBodRpll8ACLHDQAOg0QLlprNFp1XBtQPtXGdrpN7vNSQdPt29qdLtlQdNIa94b9AZjBtOrNtRNTvMYIBAAEkGFhVMcrgQAPoMHxlxxsMupMIEBTK45oBhjNNu93He4EDYMXWB1NGjM055q9DraL0EuYcs1zAQbP8gtF454G2t9tGzvunt9zWWtcqlNj7nDzOnrlL3MAJUwVFhm4IPMuOZAADV/iRJciZ%2BWfmIXgLte%2BaFpgxbri2AhbsaqZqnuxD9iwjpoLQEDro6SjAGwggQKG3qDpeaYjvKRFjOguZ/A4JCwhOXhTg8wJ/tWeBsE2G7Qc%2B26xvBs77tuUZegG6aEWq56oiJJ6ggQmAsMkBgyZiBAAJ7JIwrCfOKOoqWpzBsMc77Us8TBeEQxy0PMjRlg2cSpApmQQE%2BxziuYABsxyqGWmAAbQOpOS5ZjuZ5ob%2BW5HlltqUFts%2BhmBccykhZxBlhcpepsjuUXtiZZkAO6YJcCBcRGxwQJ53liHavrhVwpyBWFwWVcVnl6l8ABizkCcO6olWVPnHAAtNVNrXD1EUDUNoncgh/YJbVFjHHlBXPgAVCVqXjQlx5yuJpEyXJCmfI%2BBgIscrXKEZ2WoOZlnENZgi2Z4woOWdxzzn5SXMmF6DCkw%2BqvLBRFeOkRjHPgVBUNWqmHcVIEruBk1ckwCJxAQEDfQ0jqlhWVbznWqANuxXzYA6tWuWFZwFjJxB2U9AjdpgACOXh4PcCjHMK5mYEjz4CJ887HPWgiOmcW2xk5l2vckZbEIYVx2qDP2Y7OLFscqCPHNcI00IWxzhKoz784Lz5hArDQlRSeAOLQyk2kwVBU69rEPKm4umVdetzskYnFeRuZeMkam3dE6yFhABBSzLRiYI6sTAGEyraRHsvR64CeO2wosu0lYMQzpnxYLQZtVR71Ze4N4fS8nMeYHHDBq7KGsjUwADWTCvQgnxhKuqBUO3nz%2B4HxzB14haxprxxKM%2BtCoHlxDszrA9xOnnxG2uYGqGPI0sF3fctkjmBszlHf3Atx%2BfPlBDH7r4Ge76voV5HVzV7XzYkNf%2Bul3fEYP1Xqd/Wq48IDB0vnvBgDBUDPliK9YgQF2YGw7hPDSy8dTRGUrvJGiJaj9kvhzdGTAlblkrCwFWmA8YE2bL%2Ba0Yts54HBpDNS5kZ5Lyqr7EALAmCqDRtUM2g1Qxo1obnKGywqEnjVE5HO9D%2B4B2YT7AgFE2HxwLg0HUeCCHY2IbjI2hNeGZyIuPGy1NHqKViAQPKjBl4C3xoIBQjp6a9kQmzS%2BnwaDEHbG/Y6z4fKwNoYw2eX9bSL0BKmaa11whWQMTTLBEBWGHFYMkMOTsVGKyxvOCA09Z6rFNvglJTsICBOWJk1ROS2BpKYcQVYPEuxVOqdyIpytUn5KSRjLGXiFzpLiIU5JytWl5OkeU3RO1QQNmlEweOf1uKpmZFk9WEsInywsmE26ESjEOSdAAViab9dWMT1hFTcLCNeVMvawgOaw1wtBCLDg4KsWgnA1m8D8NwXgqBOD7MsNYCe6xNiHXBDwUgBBNDXNWB3JgWB4gJ2biANZuxHRcF2AATlcv4VySLJAaFZLsLgSRbkcEkLwFgEgNCWkeVoUgLyOC8AUCAS0AKOBaFWHAWASBdnJFMuQSgaA5J0DiBEDSnBVDIv6q5SQJZkDIGOFix0uxeCYHwACPAFEkj8EECIMQ7ApAyEEIoFQ6g6U6D0DlGWyROA8BuXch5gKyWcAAPKmVZTzXuArXJCpFcAMVEr/QiogB4Ll9A557G9LwWl9LSAQCQJy5I3KyAUEcqgX1wxgBYukDQQucQqVAMtdEMIjRlImt4Fm5gxBlLWuiNof4ebSCcpwgQa1DBraWqwNELwwAsS0FoFSp5pAsDsKMOIPVXaWb/DwH8DtpLwJdFMtsP5NkcWkostEGWRaPBYEtSKViFa/jEGDkoW0slZYLNAHq1YVADDAAUO%2BPAmAcrWt0hW5VwhRDiA1fe7VahLW6GxbLFA1hrD6DwNEKlsA9LsHynyUgm7OC7F2KyIFZLkhYI7f1civpTDvMsFwVkA0ADqYhaDYZvjLfDxBiBv36mpb6ghLb9TiCR1xzzN2iiwIBhOnRujOAgK4SYbQgiFgGKUcoehUjpCwVxwTBQsF8aGPEUY1Qh11FmKJmTNQeizEk4saTMw%2BiKc000NTAnvRrA2FsCQZqOD3NICS55nAPKCuFaK8VkqvW4EIG/ANywg2AoZUgcdyA3bRsoI0M9yhDDVCEAgGejy/kRroLTBgwXwi0DCxFy10W/UgETZIbUqWeV8opZWuNka/W2uEOFnKlmgiqC6M8YgZ7OAysq8geo%2BBHm8Hvaqp9ybZCvt1aSj9%2Bgv2oasJYP9AH4AQGAyAUDjBwPDA4LsNFsHUDwcyIh5DEZBvWAw9h3D%2BH9aEf6lhmjpHyOGEcMgajxGSBs36tasw9GaMKoPvAVYMDKNsEZKgTwzHVgKC%2BcZ0YciwjxdC6VyLvBDVMGNU80z5nyvkuwA1vzNnnV2bdQ5yQjozAlTeUNmwPI5Wud%2Be5/5nngVczBZQVYkLJCuUdMijQmKoNcDWfCwK/h4X6E4Hi0gBKuBEos5a8llKEgk6PaZu7PPCXEsF3V0X9LVibvSM4SQQA%3D%3D) reproduces the unintialized values. Adding eval() inside linear_interpolation produces the right value. Adding either coeff() or operator[] outside as well. 

The lazy version of linear_interpolation doesn´t seem to work like this, maybe just do what you did originally :(

Reply:
Yeah, something seems to break with the lazy addition/subtraction of y1 and y2. I kind of suspect that the auto gets set to the wrong type, since when weight is 0 the functions behave as expected.

I reset to the non-lazy version. I think it is easier to understand since it makes fewer index modifications, and it appears to be marginally faster in my benchmark.